### ISSUE TITLE
Publication fails with java.lang.StackOverflowError

### ISSUE THREAD

**Issue by xephon2** [dita-ot/dita-ot#2433]
The project is a terminology database. It has many circular dependencies (keyrefs) by design, because it is a semantic net. The publication worked (until now). It seems that I know have crossed a memory border.
- DITA-OT: `2.2.5`
- Transformation: `html5` based
- `-Xmx2048m`
- Maps: `1`
- Topics: `360`

```
keyref:
   [keyref] Reading file:/C:/Users/eike/DITA/terminologie/temp/webhelp-termbrowser/oxygen_dita_temp/Terminologie.ditamap

BUILD FAILED
C:\Users\eike\DITA-OT\dita-ot\build.xml:41: The following error occurred while executing this line:
C:\Users\eike\DITA-OT\dita-ot\plugins\org.dita.base\build_preprocess.xml:270: java.lang.StackOverflowError
    at java.util.HashMap.hash(Unknown Source)
    at java.util.HashMap.containsKey(Unknown Source)
    at java.util.Collections$UnmodifiableMap.containsKey(Unknown Source)
    at org.dita.dost.reader.KeyrefReader.resolveIntermediate(KeyrefReader.java:256)
    at org.dita.dost.reader.KeyrefReader.resolveIntermediate(KeyrefReader.java:261)
    at org.dita.dost.reader.KeyrefReader.resolveIntermediate(KeyrefReader.java:261) (this line repeats)
    ...

Total time: 12 seconds

The process finished with exit code: 1
```

I hope that `KeyrefReader.java:261` points to a processing step, that can be easily refactored. Unfortunately I cannot reproduce the issue with a test case, so I hope the provided information is enough.

**Comment by xephon2** [comment]
I found the problem in my map. I added the attribute `@keyref` by mistake additionally to `@href` and `@keys` to a `<topicref>`. I am not sure whether I should close the ticket, because I'm not sure whether the publication should work anyway. I am not sure, if I have the specs in mind correctly, but I think, this should work and pick the `@keyref` and use the `@href` as a fallback.

```xml
<topicref keyref="TERM_X" href="TERM_X.dita" keys="TERM_X"/>
```

In this case, the attributes do not lead to errors.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE map PUBLIC "-//OASIS//DTD DITA Map//EN" "map.dtd">
<map>
 <title>DITA Topic Map</title>
 <topicref keyref="topic.dita" href="topic.dita" keys="topic"/>
</map>
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE topic PUBLIC "-//OASIS//DTD DITA Topic//EN" "topic.dtd">
<topic id="topic_j5l_qpw_rw">
  <title>Foo</title>
  <body>
    <p>Bar</p>
  </body>
</topic>
```

**Comment by drmacro** [comment]
It is an error for a topicref to directly or indirectly refer to itself (no cycles are allowed) so I think the code is at fault for not detecting the cycle in this case.

**Comment by xephon2** [comment]
Thanks @drmacro. @jelovirt, I think the DITA-OT should throw a readable error message here, because the problem can be easily detected (`@keyref == @keys`). If you do not agree, please close this issue.

**Comment by drmacro** [comment]
This specific error can easily be detected but I would expect it to be detected by a more general cycle detection done during key space construction. Consider this example:

```
<topicref keys="t1" keyref="t2"/>
<topicref keys="t2" keyref="t1"/>
```

This case should also be detected. Of course it may be that a topicref that points directly to itself is a special case depending on how key space construction is implemented and when intermediate key references are resolved.

**Comment by raducoravu** [comment]
This looks like a duplicate of: https://github.com/dita-ot/dita-ot/issues/2337

**Comment by eerohele** [comment]
Or #2227, which was fixed in DITA-OT 2.3? Maybe try upgrading?

**Comment by xephon2** [comment]
@eerohele: Problem is solved, as I said, by removing the `@keyref`.
